internal/step/foreach/provider.go

```diff
@@ -441,14 +564,33 @@ func (r *runningStep) run() {
 		r.wg.Done()
 	}()
 	waitingForInput := false
+	enabled, contextDoneEarly := r.enableStage()
+	if contextDoneEarly {
+		r.closedEarly(StageIDExecute, true)
+		return
+	}
+	if !enabled {
+		r.transitionToDisabled()
+		return
+	}
+
+	var newState step.RunningStepState
 	r.lock.Lock()
-	if !r.inputAvailable {
-		r.currentState = step.RunningStepStateWaitingForInput
+	if !r.executionInputAvailable {
+		newState = step.RunningStepStateWaitingForInput
 		waitingForInput = true
 	} else {
-		r.currentState = step.RunningStepStateRunning
+		newState = step.RunningStepStateRunning
 	}
 	r.lock.Unlock()
+	enabledOutput := any(map[any]any{"enabled": true})
```

Comment:
Code like this where you unlock and then call out always scares me; I guess I'm too lazy to pull the code into an editor where I can really trace the call trees, and GitHub's limited definition/reference support is wholly inadequate to the task, so I'll simply trust that everything after the unlock is unshared, atomic, or covered by a separate lock in such a way that no broken predicates can be exposed. 🤞🏻

Reply:
transitionStageWithOutput locks on its own.

Reply:
@jaredoconnell, what scares @dbutenhof is the fact that unlocking a lock for a call-out (assuming that the calling code reacquires the lock upon return from the call) creates a "hole" in the _critical section_ that is protected by the lock.  This is not _necessarily_ a problem, and, in many cases it is the appropriate approach (or, at least, an _expedient_ solution...), but it creates an exposure to problems if the calling code (or its caller) makes assumptions about the atomicity of the critical section (particularly if such assumptions are introduced with later code changes).

This practice is especially noisome when the calling code unlocks the lock specifically so that the called code can acquire that same lock.  This just generates overhead, and it is likely to create inefficiency when the lock is under contention.  And, it "smells" of poor synchronization design.

So, Dave's concerns are warranted.  But, they are hard to allay and often hard to address.  So, mostly, his comment is just a heads-up to try to avoid this practice when you are tempted to engage in it or to remediate when you encounter it.

Reply:
Changing this would require a large refactor. I don't like the current design. A lot of it is just what we inherited. If you want to spend a part of a Friday refactoring the locking in the step providers, I'm open to that.

Reply:
Understood.  As for refactoring, that's a tempting invitation, but, for now, I'm afraid I have other demands on me.